// bold-ui/lib/components/PivotTable/Box/Box.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Box = void 0;
var react_1 = __importDefault(require("react"));
var styles_1 = require("../../../styles");
var Icon_1 = require("../../Icon");
function Box(props) {
    var styles = props.styles, icon = props.icon, rotation = props.rotation, label = props.label, children = props.children;
    var _a = styles_1.useStyles(createStyles), classes = _a.classes, css = _a.css;
    return (react_1.default.createElement("div", { className: classes.wrapper },
        react_1.default.createElement("div", { className: classes.panel },
            react_1.default.createElement("h4", { className: classes.header },
                icon && (react_1.default.createElement(Icon_1.Icon, { icon: icon, style: css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n                transform: rotate(", "deg);\n              "], ["\n                transform: rotate(", "deg);\n              "])), rotation) })),
                label)),
        react_1.default.createElement("div", { className: css(classes.children, styles) }, children)));
}
exports.Box = Box;
Box.defaultProps = {
    rotation: '0',
};
var createStyles = function (theme) { return ({
    header: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
    },
    panel: {
        backgroundColor: theme.pallete.gray.c90,
        height: '2.5rem',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        borderBottom: "1px solid " + theme.pallete.divider,
    },
    wrapper: {
        border: "1px solid " + theme.pallete.divider,
    },
    children: {
        minHeight: '7.18rem',
    },
}); };
var templateObject_1;
